### DISCUSSION TITLE
Integrate Meshtastic into an existing cpp application.

### DISCUSSION THREAD

**Discussion by sticilface** [meshtastic/firmware#4720] [General]
I'm designing an application + PCB for ESP32 that controls services in a campervan. I've also added a LORA radio to the pcb. 

Is it possible to add the Meshtastic back end so that messages can be received and routed by my application to other nodes on the network. I do not for example need the screen and Bluetooth services.

**Comment by GUVWAF** [comment]
Unfortunately that's not so easy. It's not designed as a library, but as a complete firmware on its own including "operating system".

You can, however, use a secondary MCU that talks to a Meshtastic node via the [serial module](https://meshtastic.org/docs/configuration/module/serial/), or, if your application is small and not resource intensive, it can be added as a [module](https://meshtastic.org/docs/development/device/module-api/) inside the Meshtastic firmware.

**Reply by beimoe** [reply]
I would like to do the same. Not so easy means it is possible :-)
Is there a overview somewhere where one can see the interaction of the software modules in meshtastic firmware as a starting point for me?

**Reply by GUVWAF** [reply]
The ReplyModule is an example of a module that reacts on a received packet: https://github.com/meshtastic/firmware/blob/03f69b3b77c49a4b8977aa87a22971bf81ebd6d9/src/modules/ReplyModule.cpp

The GenericThreadModule is an example of a module that runs periodically, where e.g. you can call `service->sendToMesh()` to transmit something:
<https://github.com/meshtastic/firmware/blob/03f69b3b77c49a4b8977aa87a22971bf81ebd6d9/src/modules/GenericThreadModule.cpp>
